rewrite-java/src/main/java/org/openrewrite/java/RemoveImport.java

```diff
@@ -58,48 +58,85 @@ public RemoveImport(String type, boolean force) {
         J j = tree;
         if (tree instanceof JavaSourceFile) {
             JavaSourceFile cu = (JavaSourceFile) tree;
-            ImportLayoutStyle importLayoutStyle = Optional.ofNullable(((SourceFile) cu).getStyle(ImportLayoutStyle.class))
+            boolean isKotlin = !(cu instanceof J.CompilationUnit) && (cu.getSourcePath().toString().endsWith("kt") || cu.getSourcePath().toString().endsWith("kts")); // poor man's `cu instanceof K.CompilationUnit`
+            ImportLayoutStyle importLayoutStyle = Optional.ofNullable(cu.getStyle(ImportLayoutStyle.class))
                     .orElse(IntelliJ.importLayout());
 
             boolean typeUsed = false;
+            Set<String> types = new HashSet<>(singleton(type));
             Set<String> otherTypesInPackageUsed = new TreeSet<>();
 
             Set<String> methodsAndFieldsUsed = new HashSet<>();
             Set<String> otherMethodsAndFieldsInTypeUsed = new TreeSet<>();
             Set<String> originalImports = new HashSet<>();
             for (J.Import cuImport : cu.getImports()) {
-                if (cuImport.getQualid().getType() != null) {
-                    originalImports.add(((JavaType.FullyQualified) cuImport.getQualid().getType()).getFullyQualifiedName().replace("$", "."));
+                JavaType.FullyQualified fq = TypeUtils.asFullyQualified(cuImport.getQualid().getType());
+                if (fq != null) {
+                    String fqnType = TypeUtils.toFullyQualifiedName(fq.getFullyQualifiedName());
+                    originalImports.add(fqnType);
+                    if (isKotlin && type.equals(fqnType)) {
+                        // For Kotlin, the owning class interfaces with methods can be used without actually importing those interfaces directly...
+                        JavaType.Class owningClass = TypeUtils.asClass(fq.getOwningClass());
+                        if (owningClass != null) {
+                            Queue<JavaType.FullyQualified> toVisit = new LinkedList<>(owningClass.getInterfaces());
+                            Set<JavaType.FullyQualified> visited = new HashSet<>();
+                            while (!toVisit.isEmpty()) {
+                                JavaType.FullyQualified current = toVisit.poll();
+                                if (!visited.add(current)) {
+                                    continue;
+                                }
+                                toVisit.addAll(current.getInterfaces());
+                            }
```

Comment:
Not sure if we'd prefer nested forloops here instead of a Queue

Reply:
I must've misread the code initially, thinking it could be solved with a nested forloop. Your arguments for keeping it like this are strong and valid :)